src/scheduler/scheduler.ts

```diff
@@ -0,0 +1,181 @@
+           await Promise.allSettled(this.schedTasks);
+        }
+        catch (e) {
+           // Throws do not matter
+        }
+        this.schedTasks = [];
+    }
+  
+    logRegisteredSchedulerEndpoints() {
+        const logger = this.dbosExec.logger;
+        logger.info("Scheduled endpoints:");
+        this.dbosExec.registeredOperations.forEach((registeredOperation) => {
+            const ro = registeredOperation as SchedulerRegistration<unknown, unknown[], unknown>;
+            if (ro.schedulerConfig) {
+                logger.info(`    ${ro.name} @ ${ro.schedulerConfig.crontab}; ${ro.schedulerConfig.mode ?? 'Exactly Once Per Interval'}`);
+            }
+        });
+    }
+}
+
+class DetachableLoop {
+    private isRunning: boolean = false;
+    private resolveCompletion?: (value: void | PromiseLike<void>) => void;
+    private interruptResolve?: () => void;
+    private lastExec: Date;
+    private timeMatcher: TimeMatcher;
+
+    constructor(readonly dbosExec: DBOSExecutor, readonly crontab: string, readonly scheduledMethod: SchedulerRegistration<unknown, unknown[], unknown>) {
+        this.lastExec = new Date();
+        this.lastExec.setMilliseconds(0);
+        this.timeMatcher = new TimeMatcher(crontab);
+    }
+
+    async startLoop(): Promise<void> {
+        // See if the exec time is available in durable storage...
+        const lasttm = await this.dbosExec.systemDatabase.getLastScheduledTime(this.scheduledMethod.name);
+        if (lasttm) {
+            this.lastExec = new Date(lasttm);
+        }
+
+        this.isRunning = true;
+        while (this.isRunning) {
+            const nextExecTime = this.timeMatcher.nextWakeupTime(this.lastExec);
+            const sleepTime = nextExecTime.getTime() - new Date().getTime();
+
+            if (sleepTime > 0) {
+                // Wait for either the timeout or an interruption
+                await Promise.race([
+                    this.sleep(sleepTime),
+                    new Promise<void>((_, reject) => this.interruptResolve = reject)
+                ])
+                .catch(); // Interrupt sleep throws
+            }
+
+            if (!this.isRunning) {
+                break;
+            }
+
+            // Check crontab
+            if (!this.timeMatcher.match(nextExecTime)) {
+                this.lastExec = nextExecTime;
+                continue;
+            }
+
+            // Init workflow
+            const workflowUUID = `sched-${this.scheduledMethod.name}-${nextExecTime.toISOString()}`;
+            this.dbosExec.logger.debug(`Executing scheduled workflow ${workflowUUID}`);
+            const wfParams = { workflowUUID: workflowUUID };
+            // All operations annotated with Scheduled decorators must take in these four
+            const args: ScheduledArgs = [nextExecTime, new Date(), 0, 0]; // TODO calculate outstanding numbers
+
+            // We can only guarantee exactly-once-per-message execution of transactions and workflows.
```

Comment:
Seems that we currently only support workflows
```suggestion
            // We currently only support scheduled workflows
```